File: swift/Sources/AttomusOTP/OTPValidation.swift
import Foundation

enum OTPValidation {
    static let minimumSecretLength = 16
    static let maximumSecretLength = 64
    static let validDigits: Set<Int> = [6, 7, 8]
    static let validPeriods: Set<Int> = [30, 60]

    static func validateHOTPSecret(_ secret: Data) throws {
        try validateSecret(secret, shortError: HOTPError.secretTooShort, longError: HOTPError.secretTooLong)
    }

    static func validateTOTPSecret(_ secret: Data) throws {
        try validateSecret(secret, shortError: TOTPError.secretTooShort, longError: TOTPError.secretTooLong)
    }

    static func validateHOTPDigits(_ digits: Int) throws {
        guard validDigits.contains(digits) else {
            throw HOTPError.invalidDigitCount
        }
    }

    static func validateTOTPDigits(_ digits: Int) throws {
        guard validDigits.contains(digits) else {
            throw TOTPError.invalidDigitCount
        }
    }

    static func validateTOTPPeriod(_ period: Int) throws {
        guard validPeriods.contains(period) else {
            throw TOTPError.invalidPeriod
        }
    }

    static func validateURISecret(_ secret: Data) throws {
        guard secret.count >= minimumSecretLength else {
            throw OTPURIError.secretTooShort
        }

        guard secret.count <= maximumSecretLength else {
            throw OTPURIError.secretTooLong
        }
    }

    static func validateURIDigits(_ digits: Int) throws {
        guard validDigits.contains(digits) else {
            throw OTPURIError.invalidDigits
        }
    }

    static func validateURIPeriod(_ period: Int) throws {
        guard validPeriods.contains(period) else {
            throw OTPURIError.invalidPeriod
        }
    }

    private static func validateSecret<E: Error>(_ secret: Data, shortError: E, longError: E) throws {
        guard !secret.isEmpty else {
            throw shortError
        }

        guard secret.count >= minimumSecretLength else {
            throw shortError
        }

        guard secret.count <= maximumSecretLength else {
            throw longError
        }
    }
}

File: kotlin/src/main/kotlin/com/attomus/otp/OTPValidation.kt
package com.attomus.otp

internal object OTPValidation {
    const val minimumSecretLength = 10
    const val maximumSecretLength = 64

    private val validDigits = setOf(6, 7, 8)
    private val validPeriods = setOf(30, 60)

    fun validateHOTPSecret(secret: ByteArray) {
        validateSecret(
            secret = secret,
            shortError = HOTPError.SecretTooShort,
            longError = HOTPError.SecretTooLong
        )
    }

    fun validateTOTPSecret(secret: ByteArray) {
        validateSecret(
            secret = secret,
            shortError = TOTPError.SecretTooShort,
            longError = TOTPError.SecretTooLong
        )
    }

    fun validateHOTPDigits(digits: Int) {
        if (digits !in validDigits) {
            throw HOTPError.InvalidDigitCount
        }
    }

    fun validateTOTPDigits(digits: Int) {
        if (digits !in validDigits) {
            throw TOTPError.InvalidDigitCount
        }
    }

    fun validateTOTPPeriod(period: Int) {
        if (period !in validPeriods) {
            throw TOTPError.InvalidPeriod
        }
    }

    fun validateURISecret(secret: ByteArray) {
        when {
            secret.size < minimumSecretLength -> throw OTPURIError.SecretTooShort
            secret.size > maximumSecretLength -> throw OTPURIError.SecretTooLong
        }
    }

    fun validateURIDigits(digits: Int) {
        if (digits !in validDigits) {
            throw OTPURIError.InvalidDigits
        }
    }

    fun validateURIPeriod(period: Int) {
        if (period !in validPeriods) {
            throw OTPURIError.InvalidPeriod
        }
    }

    @Suppress("ThrowsCount")
    private fun validateSecret(
        secret: ByteArray,
        shortError: Exception,
        longError: Exception
    ) {
        when {
            secret.isEmpty() -> throw shortError
            secret.size < minimumSecretLength -> throw shortError
            secret.size > maximumSecretLength -> throw longError
        }
    }
}
